buffer/src/main/java/io/netty/buffer/InstrumentedByteBufAllocator.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer;
+
+/**
+ * {@link ByteBufAllocator} which exposes metrics.
+ */
+public interface InstrumentedByteBufAllocator extends ByteBufAllocator {
```

Comment:
+1 for `ByteBufAllocatorMetrics` but I guess a user might expect it to be something like this:

```java
public interface ByteBufAllocatorMetricsProvider {
    ByteBufAllocatorMetrics metrics();
}

public class UnpooledByteBufAllocator extends ... implementa ByteBufAllocatorMetricsProvider {
    ...
}

alloc.metrics().usedHeapMemory();
```

Given that I do not have time to implement this proposal, please consider this just as a suggestion. It's you guys' right to choose what you prefer.